File: speech_recognition_demo/python/speech_recognition_demo/speech_recognition_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Speech Recognition Demo Node (Python version)

订阅 /lyre/voice_activity 话题获取ASR语音活动事件。
统一接收人脸唤醒、关键词唤醒、VAD事件、ASR识别结果等。
"""

import json

import rclpy
from rclpy.node import Node
from lyre_msgs.msg import LyreVoiceActivity


EVENT_TYPE_NAMES = {
    1: 'ASR结果',
    4: '关键词唤醒',
    5: '退出对话',
    6: 'VAD事件',
    20: '人脸识别唤醒',
}


class SpeechRecognitionNode(Node):
    """ROS2 Node for speech recognition demo via /lyre/voice_activity"""

    def __init__(self):
        super().__init__('speech_recognition_demo')

        self.voice_activity_sub = self.create_subscription(
            LyreVoiceActivity,
            '/lyre/voice_activity',
            self.voice_activity_callback,
            10
        )

        self.get_logger().info('SpeechRecognitionNode initialized')
        self.get_logger().info('Subscribed to: /lyre/voice_activity')

    def voice_activity_callback(self, msg):
        """Parse and display voice activity events"""
        try:
            data = json.loads(msg.content)
        except json.JSONDecodeError:
            self.get_logger().warn(f'Failed to parse JSON: {msg.content[:200]}')
            return

        event_type = data.get('type', '')
        content = data.get('content', {})
        trace_id = data.get('traceId', '')
        timestamp = data.get('timestamp', '')

        if event_type == 'aiui_event':
            self._handle_aiui_event(content, trace_id, timestamp)
        else:
            self.get_logger().info(f'[VoiceActivity] type={event_type}, content={msg.content[:200]}')

    def _handle_aiui_event(self, content, trace_id, timestamp):
        event_type = content.get('eventType', -1)
        event_name = EVENT_TYPE_NAMES.get(event_type, f'未知事件({event_type})')

        if event_type == 1:
            result = content.get('result', {})
            text_data = result.get('text', {})
            ws_list = text_data.get('ws', [])
            recognized_text = ''
            for ws in ws_list:
                cw_list = ws.get('cw', [])
                for cw in cw_list:
                    recognized_text += cw.get('w', '')
            self.get_logger().info(
                f'[ASR] 识别结果: "{recognized_text}" (traceId={trace_id})')

        elif event_type == 4:
            result = content.get('result', {})
            ivw = result.get('ivw', {})
            angle = ivw.get('angle', -1)
            self.get_logger().info(
                f'[唤醒] 关键词唤醒，角度: {angle}° (traceId={trace_id})')

        elif event_type == 5:
            self.get_logger().info(f'[对话] 退出对话 (traceId={trace_id})')

        elif event_type == 6:
            arg1 = content.get('arg1', -1)
            if arg1 == 0:
                self.get_logger().info(f'[VAD] 检测到语音开始 (traceId={trace_id})')
            elif arg1 == 2:
                self.get_logger().info(f'[VAD] 检测到语音结束 (traceId={trace_id})')
            else:
                self.get_logger().info(
                    f'[VAD] VAD事件 arg1={arg1} (traceId={trace_id})')

        elif event_type == 20:
            self.get_logger().info(f'[唤醒] 人脸识别唤醒 (traceId={trace_id})')

        else:
            self.get_logger().info(
                f'[事件] {event_name}, eventType={event_type} (traceId={trace_id})')


def main(args=None):
    rclpy.init(args=args)
    node = SpeechRecognitionNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: 4.speech_recognition_demo/cpp/speech_recognition_demo/src/speech_recognition_node.cpp
/**
 * @file speech_recognition_node.cpp
 * @brief 订阅 /lyre/voice_activity 话题获取ASR语音活动事件
 *
 * 统一接收人脸唤醒、关键词唤醒、VAD事件、ASR识别结果等
 */

#include <memory>
#include <string>
#include <sstream>

#include "lyre_msgs/msg/lyre_voice_activity.hpp"
#include "rclcpp/rclcpp.hpp"

// Minimal JSON value extraction helpers (avoid external JSON lib dependency)
static std::string extract_string(const std::string& json, const std::string& key) {
  std::string search = "\"" + key + "\":\"";
  auto pos = json.find(search);
  if (pos == std::string::npos) return "";
  pos += search.size();
  auto end = json.find('"', pos);
  if (end == std::string::npos) return "";
  return json.substr(pos, end - pos);
}

static int extract_int(const std::string& json, const std::string& key, int def = -1) {
  std::string search = "\"" + key + "\":";
  auto pos = json.find(search);
  if (pos == std::string::npos) return def;
  pos += search.size();
  try { return std::stoi(json.substr(pos)); } catch (...) { return def; }
}

static std::string extract_asr_text(const std::string& json) {
  std::string result;
  std::string search = "\"w\":\"";
  size_t pos = 0;
  while ((pos = json.find(search, pos)) != std::string::npos) {
    pos += search.size();
    auto end = json.find('"', pos);
    if (end != std::string::npos) {
      result += json.substr(pos, end - pos);
      pos = end + 1;
    }
  }
  return result;
}

class SpeechRecognitionNode : public rclcpp::Node {
 public:
  SpeechRecognitionNode() : Node("speech_recognition_demo") {
    voice_activity_sub_ = this->create_subscription<lyre_msgs::msg::LyreVoiceActivity>(
        "/lyre/voice_activity", 10,
        std::bind(&SpeechRecognitionNode::voice_activity_callback, this,
                  std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "SpeechRecognitionNode initialized");
    RCLCPP_INFO(this->get_logger(), "Subscribed to: /lyre/voice_activity");
  }

 private:
  void voice_activity_callback(const lyre_msgs::msg::LyreVoiceActivity::SharedPtr msg) {
    const auto& content = msg->content;
    std::string type = extract_string(content, "type");
    std::string trace_id = extract_string(content, "traceId");

    if (type != "aiui_event") {
      RCLCPP_INFO(this->get_logger(), "[VoiceActivity] type=%s", type.c_str());
      return;
    }

    int event_type = extract_int(content, "eventType");

    switch (event_type) {
      case 1: {
        std::string text = extract_asr_text(content);
        RCLCPP_INFO(this->get_logger(), "[ASR] 识别结果: \"%s\" (traceId=%s)",
                    text.c_str(), trace_id.c_str());
        break;
      }
      case 4: {
        int angle = extract_int(content, "angle");
        RCLCPP_INFO(this->get_logger(), "[唤醒] 关键词唤醒，角度: %d° (traceId=%s)",
                    angle, trace_id.c_str());
        break;
      }
      case 5:
        RCLCPP_INFO(this->get_logger(), "[对话] 退出对话 (traceId=%s)",
                    trace_id.c_str());
        break;
      case 6: {
        int arg1 = extract_int(content, "arg1");
        if (arg1 == 0) {
          RCLCPP_INFO(this->get_logger(), "[VAD] 检测到语音开始 (traceId=%s)",
                      trace_id.c_str());
        } else if (arg1 == 2) {
          RCLCPP_INFO(this->get_logger(), "[VAD] 检测到语音结束 (traceId=%s)",
                      trace_id.c_str());
        } else {
          RCLCPP_INFO(this->get_logger(), "[VAD] VAD事件 arg1=%d (traceId=%s)",
                      arg1, trace_id.c_str());
        }
        break;
      }
      case 20:
        RCLCPP_INFO(this->get_logger(), "[唤醒] 人脸识别唤醒 (traceId=%s)",
                    trace_id.c_str());
        break;
      default:
        RCLCPP_INFO(this->get_logger(), "[事件] eventType=%d (traceId=%s)",
                    event_type, trace_id.c_str());
        break;
    }
  }

  rclcpp::Subscription<lyre_msgs::msg::LyreVoiceActivity>::SharedPtr voice_activity_sub_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SpeechRecognitionNode>());
  rclcpp::shutdown();
  return 0;
}
